Sources/NIOHPACK/HPACKHeader.swift

```diff
@@ -0,0 +1,443 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+/// Very similar to `NIOHTTP1.HTTPHeaders`, but we have our own version to make some
+/// small changes inside.
+public struct HPACKHeaders {
```

Comment:
The `NIOHTTP1.HTTPHeaders` structure will need to be reworked a little to enable this without round-tripping every header name & value via `String`. As it stands, all API regarding the byte buffer and header indices is `internal`. Ideally we'd want to be able to transfer easily between the two by just taking a copy of the `ByteBuffer` directly and mapping between `HTTPHeader` and `HPACKHeader` collections, but right now that's not possible. Until then, this would need basically the same API as `HTTPHeaders`, no?

Reply:
Agreed that `HTTPHeaders` needs to be reworked to support this, but that's fine, we can do that when needed. It's not necessary right now because we aren't plugging this into the top-level NIOHTTP2 API yet, and so it doesn't need to actually bridge into that data structure.

More importantly, though, I want to avoid committing to APIs we don't need at this early stage. Just because we can bring over the API that `HTTPHeaders` has doesn't mean we should: its API is large and it does a bunch of stuff. I'd be very inclined to start by having this structure be very small: a collection of header names and values would be just fine. Everything else can be added later if users want it.

As a general design principle it is much easier to add APIs than to take them away. I'd rather start small and have users complain that things are missing than to start big and then have to support the whole thing for three years.

Reply:
So what would the API be on `HPACKHeaders`? Just creation and enumeration? Would it exist then purely as a transformation on `HTTPHeaders`? If we want to enable things like proxy rewriting at a later time we'd need to be able to support adding or replacing headers along with their indexing behavior; if this type only exists as a translation layer between `HTTPHeaders` and the HPACK encoders, then we're losing something there.

Either way, I'll pull the add/remove methods for now, but I suspect we'll need to pull them back at some point; they're in the commit history, so that should help.

Reply:
So I've removed the remove/rename methods here and made the additions private (they're used by constructors), but I've left the subscript getters in place.